pandas/_libs/src/vendored/numpy/datetime/np_datetime.c

```diff
@@ -483,14 +483,27 @@ npy_datetime npy_datetimestruct_to_datetime(NPY_DATETIMEUNIT base,
   if (base == NPY_FR_ns) {
     int64_t nanoseconds;
 
-    // for near-minimum timestamps, scaling microseconds to nanoseconds
-    // overflows but adding nanoseconds puts the timestamp back in a valid range
+    // Minimum valid timestamp in nanoseconds (1677-09-21 00:12:43.145224193).
     const int64_t min_nanoseconds = NPY_MIN_INT64 + 1;
+
+    // For near-minimum timestamps (1677-09-21 00:12:43.145224193 through
+    // 1677-09-21 00:12:43.145224999), scaling microseconds to nanoseconds
+    // overflows (1677-09-21 00:12:43.145224 -> 1677-09-21 00:12:43.145224000),
+    // but adding nanoseconds can put the timestamp back in a valid range for
+    // nanosecond parts >= 193.
+
+    // (min_nanoseconds / 1000 - 1) * 1000 would overflow, so do not scale.
+    // This happens if microseconds corresponds to 1677-09-21 00:12:43.145224.
     if (microseconds == min_nanoseconds / 1000 - 1) {
```

Comment:
Can you check performance on this versus main and just make sure no regression? I would hope it doesn't make a difference but this is in a really tight loop so small differences can add up.

Reply:
It seems like there is no measurable performance difference between the two.

Current main (2493681):
```bash
$ python -m timeit --setup "import numpy as np; from pandas._libs import tslib" "tslib.array_to_datetime(np.array(['1677-09-21T00:12:43.145225000'], dtype=object))"
200000 loops, best of 5: 1.33 usec per loop
```

This branch ([9c7c0d3](https://github.com/robert-schmidtke/pandas/tree/cbb9280fb094640a66b97026cb0ca3a909a59b5d)):
```bash
$ python -m timeit --setup "import numpy as np; from pandas._libs import tslib" "tslib.array_to_datetime(np.array(['1677-09-21T00:12:43.145225000'], dtype=object))"
200000 loops, best of 5: 1.3 usec per loop

# testing the added branch for near-minimum timestamp handling
$ python -m timeit --setup "import numpy as np; from pandas._libs import tslib" "tslib.array_to_datetime(np.array(['1677-09-21T00:12:43.145224200'], dtype=object))"
200000 loops, best of 5: 1.31 usec per loop
```